xds/src/main/java/io/grpc/xds/internal/rbac/engine/AuthorizationEngine.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.rbac.engine;
+
+import com.google.api.expr.v1alpha1.Expr;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.Descriptors.Descriptor;
+import io.envoyproxy.envoy.config.rbac.v2.Policy;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC;
+import io.envoyproxy.envoy.config.rbac.v2.RBAC.Action;
+import io.grpc.xds.internal.rbac.engine.cel.Activation;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultDispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.DefaultInterpreter;
+import io.grpc.xds.internal.rbac.engine.cel.DescriptorMessageProvider;
+import io.grpc.xds.internal.rbac.engine.cel.Dispatcher;
+import io.grpc.xds.internal.rbac.engine.cel.IncompleteData;
+import io.grpc.xds.internal.rbac.engine.cel.Interpretable;
+import io.grpc.xds.internal.rbac.engine.cel.Interpreter;
+import io.grpc.xds.internal.rbac.engine.cel.InterpreterException;
+import io.grpc.xds.internal.rbac.engine.cel.RuntimeTypeProvider;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * CEL Evaluation Engine is part of the authorization framework in gRPC.
+ * CEL Evaluation Engine takes one or two Envoy RBAC policies as input
+ * and uses CEL library to evaluate the condition field 
+ * inside each RBAC policy based on the provided Envoy Attributes. 
+ * CEL Evaluation Engine will generate an authorization decision which
+ * could be ALLOW, DENY or UNKNOWN.
+ * 
+ * <p>Use as in:
+ * 
+ * <pre>
+ *  AuthorizationEngine engine = new AuthorizationEngine(
+ *      ImmutableList.of(rbacPolicies));
+ *  AuthorizationDecision result = engine.evaluate(new EvaluateArgs(call, headers));
+ * </pre>
+ */
+public class AuthorizationEngine<ReqT, RespT> {
```

Comment:
Can it be `final` ?

Reply:
Similarly as above, using Mockito for unit testing has a lot of limitations including not able to mock a final class. However, this is definitely something that can be changed in the future :)